src/http.c

```diff
@@ -534,11 +536,18 @@ int auth_log_in(struct tunnel *tunnel)
 		get_value_from_response(res, "reqid=", reqid, 32);
 		get_value_from_response(res, "polid=", polid, 32);
 
-		read_password("Two-factor authentication token: ", tokenresponse, 255);
+		if (cfg->otp[0] == '\0') {
+			read_password("Two-factor authentication token: ",
+			              cfg->otp, FIELD_SIZE);
+			if (cfg->otp[0] == '\0') {
+				log_error("No token specified\n");
+				return 0;
+			}
+		}
 
 		snprintf(data, 256, "username=%s&realm=%s&reqid=%s&polid=%s&grp=%s"
 		         "&code=%s&code2=&redir=%%2Fremote%%2Findex&just_logged_in=1",
-		         username, realm, reqid, polid, group, tokenresponse);
+		         username, realm, reqid, polid, group, cfg->otp);
```

Comment:
actually, this was my initial concern. But this one-time password that's read from a file somewhere in /dev/shm, isn't it something the nm-applet queries from the user, i.e. something like a token, a PIN sent via SMS or so?

Reply:
on the other hand, in the current code there is an *urlencode()* on *tokenresponse*, so removing this might break things for some users and adding it to the new code wouldn't modify anything for tokens that don't contain special characters...